# Challenge 2 

## Object Tracker By Color

Build Object Tracker By Color 

Function will ask the user about the object to  track

User feed string input  e.g. "Blue Shirt"

Function would use the first word to choose  the corresponding lower and upper limit in range method 

Make the function ready for at least 3 colors ( Red , Green , Blue )

In [1]:
import cv2
import numpy as np

In [4]:
def lower_upper(color_name):
    # define range of blue color in HSV
    lower_blue = np.array([105,50,50])
    upper_blue = np.array([130,255,255])

    # define range of green color in HSV
    lower_green = np.array([45,50,50])
    upper_green = np.array([75,255,255])

    # define range of red color in HSV
    lower_red = np.array([0,50,50])
    upper_red = np.array([10,255,255])
    
    # Color Testing
    ct = color_name[0:3]
    if ct.lower()=='blu':
        lower = lower_blue
        upper = upper_blue
    elif ct.lower()=='gre':
        lower = lower_green
        upper = upper_green
    elif ct.lower()=='red':
        lower = lower_red
        upper = upper_red
    else:
        # User Input is not correct
        lower = np.array([0,0,0])
        upper = np.array([0,0,0])
    
    return lower, upper, ct.lower();

In [8]:
def object_tracker():
    
    while(1):
        user_color = input('Kindly write object name to track it: ')
        lower, upper, ct = lower_upper(user_color)
        if lower[1] != 0:
            print('Lower Range: (' + str(lower[0]) + ', ' + str(lower[1]) + ', ' + str(lower[2]) + ')')
            print('Upper Range: (' + str(upper[0]) + ', ' + str(upper[1]) + ', ' + str(upper[2]) + ')')
            break

    while(1):
        user_video = input('Write your selection (Video / Live): ')
        if user_video.lower() == 'video':
            # Select one of the following 2 code lines for: (HDD Video ,or Online Video)
            # Video Stored in your HDD
            cap = cv2.VideoCapture('Smurfs.mp4')
            # Online Video
            #cap = cv2.VideoCapture('http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4')

            break
        elif user_video.lower() == 'live':
            cap = cv2.VideoCapture(0)
            break


    while(1):
        x, frame = cap.read()
        hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

        # Threshold the HSV image to get only blue colors
        if ct == 'red':
            lower_2 = np.array([170,50,50])
            upper_2 = np.array([179,255,255])
            mask_1 = cv2.inRange(hsv, lower, upper)
            mask_2 = cv2.inRange(hsv, lower_2, upper_2)
            mask = mask_1 + mask_2
        else:
            mask = cv2.inRange(hsv, lower, upper)

        res = cv2.bitwise_and(frame,frame, mask= mask)

        cv2.imshow('frame',frame)
        cv2.imshow('mask',mask)
        cv2.imshow('res',res)

        if cv2.waitKey(10) & 0xFF == ord('q'):
            break

    cv2.destroyAllWindows()
    cap.release()

In [7]:
# Call Main Function
object_tracker()

Kindly write object name to track it:  Blue Smurfs


Lower Range: (105, 50, 50)
Upper Range: (130, 255, 255)


Write your selection (Video / Live) Video


# Helping & Testing Code:

In [9]:
# To Determine Color Value in HSV

blue = np.uint8([[[255,0,0]]])
hsv_blue = cv2.cvtColor(blue, cv2.COLOR_BGR2HSV)
print('HSV of Blue: ' + str(hsv_blue))

green = np.uint8([[[0,255,0 ]]])
hsv_green = cv2.cvtColor(green, cv2.COLOR_BGR2HSV)
print('HSV of Green: ' + str(hsv_green) )

red = np.uint8([[[0,0,255]]])
hsv_red = cv2.cvtColor(red, cv2.COLOR_BGR2HSV)
print('HSV of Red: ' + str(hsv_red))


HSV of Blue: [[[120 255 255]]]
HSV of Green: [[[ 60 255 255]]]
HSV of Red: [[[  0 255 255]]]


### H&deg; Color Range in HSV Color Space:

The below image shows the H&deg; color range in HSV Color Space.\
Note that: the image show 360&deg; but OpenCV uses 180&deg;, so you should divide any color degree by 2.\
For Example:
+ If you need green color (120&deg; from image),
+ you'll divide it by 2,
+ and write 60&deg; in OpenCV H value of HSV range.

![alt text](hsvColorRange.png "HSV Color Range")

# References:

+ Electro Pi Computer Vision offline Course at: https://electro-pi.com/
+ Free online mp4 Videos: https://gist.github.com/jsturgis/3b19447b304616f18657